### PR TITLE
Added a link checker step in the doctests jobs using both Lychee and …

### PR DESCRIPTION
**Pull request by RohitP2005**
# Sphinx Link Checker Integration in CI

## Description

This PR introduces a fix to run the Sphinx link checker in parallel with Lychee during the CI pipeline. This resolves the issue of missing or broken links not being detected while ensuring proper integration with the existing workflow. The Sphinx link checker builder has been added to the CI process to ensure links are checked effectively, with broken links automatically identified and flagged.

**Fixes:** #3387

---

## Type of change

- [ ] New feature (non-breaking change that adds functionality)
- [x] Optimization (back-end change that speeds up the code)
- [x] Bug fix (non-breaking change that fixes an issue)

---

## Key checklist:

- [ ] No style issues: `$ pre-commit run` (or `$ nox -s pre-commit`) (see [CONTRIBUTING.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally)
- [x] All tests pass: `$ python -m pytest` (or `$ nox -s tests`)
- [x] The documentation builds: `$ python -m pytest --doctest-plus src` (or `$ nox -s doctests`)

---

## Further checks:

- [ ] Code is commented, particularly in hard-to-understand areas
- [ ] Tests added that prove fix is effective or that feature works

---

### CODE REVIEW

File: .github/workflows/links.yml
Comment (kratman) [inline .github/workflows/links.yml:77]:
```suggestion
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}

```

Comment (kratman) [inline .github/workflows/links.yml:77]: The failure is probably coming from needing to install PyBaMM[all,docs] before running the check. I have no looked at the linkcheck option before, but you might also need to build the docs independently

Comment (kratman) [reply .github/workflows/links.yml:77]: This might be something to include int the docs build in nox instead of here. Otherwise we will be doing 3 documentation builds

Comment (RohitP2005) [reply .github/workflows/links.yml]: Thanks @kratman ,  I will make sure documentation is built before the link check

Comment (agriyakhetarpal) [reply .github/workflows/links.yml]: There are two options:
- we should use `scientific-python/cookie`'s pattern here: https://github.com/scientific-python/cookie/blob/95e0bb06ccf09163bed4e2c6d5726e5990c70a13/%7B%7Bcookiecutter.project_name%7D%7D/noxfile.py#L48-L75 and create a small argument parser that allows choosing between the HTML builder or the linkcheck builder
- or, just install PyBaMM here with `[docs,all]` without modifying the `nox` session before running `sphinx-build -b linkcheck`. The linkcheck builder calls the HTML builder internally

I slightly lean towards the second option, because I don't think anyone is going to use this session for local development.

I agree on not having redundant CI jobs, though; it'd probably be nice to move all docs-related jobs (docs build, doctests, and this linkchecker build) to a `docs.yml` workflow file. Our CI is not winning any timing competition, so they can all be steps that run in serial in a single job instead of separate jobs.

Comment (RohitP2005) [reply .github/workflows/links.yml]: @agriyakhetarpal  I went with the second option of installing PyBaMM with ```[docs,all]```  . Kindly review the new code for further modifications if needed. I also ran the pre-commit this time

**Review by kratman — CHANGES_REQUESTED** [review]
I still think this should be part of the doctests or docs workflow instead of with lychee. Lychee is really unstable so most people will not even look at the failure

**Comment by RohitP2005** [comment]
> I still think this should be part of the doctests or docs workflow instead of with lychee. Lychee is really unstable so most people will not even look at the failure

can u kindly be specific , cause it was asked to run Lychee and and Sphinx parallelly

**Comment by RohitP2005** [comment]
> Sure, what Eric meant was to move the job you added + the current docs-related tests in `test_on_push.yml` to a separate workflow file (the latter part of my previous comment)? To avoid building the docs thrice, we can build them without the `nox` sessions because a CI runner is isolated enough.

Understood  @agriyakhetarpal  i will look into it asap
Thanks for the clarification

File: .github/workflows/docs.yml
Comment (kratman) [inline .github/workflows/docs.yml:73]: @RohitP2005 I think there was a misunderstanding here. Lychee should stay in its own check. The reason is that it often fails due to connectivity issues.

Splitting gives us the following:
1. If Lychee fails and no URLs were added a developer can safely assume they did not break anything. The failure was likely networking. We cannot make an unstable check required
2. Documentation can break due to updated doc strings, the doctests (testing code blocks in the documentation), or bad links inside the documentation. If this fails it must be checked and we can easily force that to be a required check

I would say that the `nox -s doctests` job should also be put in the "test_docs_build.yml" workflow

File: .github/workflows/docs.yml
Comment (kratman) [inline .github/workflows/docs.yml:1]: Can we change this file name to `test_docs_build.yml`

File: .github/workflows/test_on_push.yml
Comment (kratman) [inline .github/workflows/test_on_push.yml:177]: This should not be commented out

**Review by kratman — CHANGES_REQUESTED** [review]

File: .github/workflows/test_on_push.yml
Comment (kratman) [reply .github/workflows/test_on_push.yml:177]: Yeah lychee can basically be moved back to the original file (just reverting the changes to that file). The `nox -s doctests` part of the commented code can be put in the `test_docs_build.yml` file. Anything redundant can be removed

File: .github/workflows/test_docs_build.yml
Comment (kratman) [inline .github/workflows/test_docs_build.yml:109]: Please just put this back in the `lychee_url_checker.yml`

File: .github/workflows/test_docs_build.yml
Comment (kratman) [inline .github/workflows/test_docs_build.yml:68]: This looks good. Can you remove the equivalent portion from the `test_on_push.yml` file?

File: .github/workflows/test_docs_build.yml
Comment (kratman) [inline .github/workflows/test_docs_build.yml:70]: Is this token needed here?

**Review by kratman — CHANGES_REQUESTED** [review]

File: .github/workflows/docs.yml
Comment (agriyakhetarpal) [inline .github/workflows/docs.yml:56]:
```suggestion
      - name: Install PyBaMM with [docs,dev,all]
        run: |
          python -m pip install --upgrade pip
          pip install -e .[docs,dev,all]
```

File: .github/workflows/docs.yml
Comment (agriyakhetarpal) [inline .github/workflows/docs.yml:64]:
```suggestion
      - name: Build docs (HTML)
        run: sphinx-build -b html -d docs/_build/.doctrees docs docs/_build
```

Building the doctrees with the HTML builder here will mean that they can be reused for the linkcheck builder in the next step.

File: .github/workflows/docs.yml
Comment (agriyakhetarpal) [inline .github/workflows/docs.yml:66]:
```suggestion
```

File: .github/workflows/test_on_push.yml
Comment (agriyakhetarpal) [reply .github/workflows/test_on_push.yml:177]: Once done, these lines can be removed entirely instead of leaving them commented out.

File: .github/workflows/docs.yml
Comment (agriyakhetarpal) [inline .github/workflows/docs.yml:68]:
```suggestion
        run: sphinx-build -b linkcheck -d docs/_build/.doctrees docs docs/_build
```

**Comment by agriyakhetarpal** [comment]
Sorry – I think some of my requests for changes might have overlapped with Eric's. Please ignore the ones that have already been done.

**Comment by kratman** [comment]
@RohitP2005 Please do not force push after the review has started. It makes it extremely hard to track changes

**Comment by RohitP2005** [comment]
> @RohitP2005 Please do not force push after the review has started. It makes it extremely hard to track changes

Sorry for the trouble eric, I will make a commit including the changes as u requested

**Comment by RohitP2005** [comment]
### Changes I Made:

- **Added** `lychee_url_checker.yml` — requested by Eric
- **Removed** the commented lines from `test_on_push.yml` — requested by @agriyakhetarpal
- **Removed** `env` and the respective token — requested by @agriyakhetarpal and Eric
- **Renamed** `docs.yml` to `test_docs_build.yml` — requested by Eric
- **Removed** the `lychee` job from the `test_docs_build.yml` — requested by Eric
- **Removed** unnecessary comments — requested by @agriyakhetarpal

File: .github/workflows/test_docs_build.yml
Comment (kratman) [inline .github/workflows/test_docs_build.yml:12]: I would say we can simply this
```suggestion
on:
  workflow_dispatch:
  pull_request:
```

Comment (RohitP2005) [reply .github/workflows/test_docs_build.yml:12]: Understood  Eric
, is there any further change ?

File: .github/workflows/test_docs_build.yml
Comment (kratman) [inline .github/workflows/test_docs_build.yml:53]: It looks like this was missed from @agriyakhetarpal's review
```suggestion
      - name: Build docs (HTML)
        run: sphinx-build -b html -d docs/_build/.doctrees docs docs/_build
```

File: .github/workflows/docs.yml
Comment (kratman) [reply .github/workflows/docs.yml]: Sorry resolved this by mistake. I re-added the request

File: .github/workflows/test_docs_build.yml
Comment (kratman) [reply .github/workflows/test_docs_build.yml]: I saw one more thing that was missed from Agriya's review,  but I think that is it from my side

File: .github/workflows/test_docs_build.yml
Comment (RohitP2005) [reply .github/workflows/test_docs_build.yml]: It was attented

File: .github/workflows/test_docs_build.yml
Comment (kratman) [reply .github/workflows/test_docs_build.yml:46]: It looks like this is still not changed when I view the review. I will wait for @agriyakhetarpal to do his last review

File: .github/workflows/test_docs_build.yml
Comment (agriyakhetarpal) [inline .github/workflows/test_docs_build.yml:43]:
```suggestion
      - name: Run doctests
        run: python -m pytest --doctest-plus src
```

I think this was missed from https://github.com/pybamm-team/PyBaMM/pull/4686#discussion_r1890504705.

File: .github/workflows/test_on_push.yml
Comment (agriyakhetarpal) [inline .github/workflows/test_on_push.yml:133]:
```suggestion
```

**Review by agriyakhetarpal — APPROVED** [review]
Just two more changes, thanks for taking this up, @RohitP2005!

**Comment by agriyakhetarpal** [comment]
@all-contributors please add @RohitP2005 for infra

**Comment by agriyakhetarpal** [comment]
Please feel free to fix the [failing links](https://github.com/pybamm-team/PyBaMM/actions/runs/12399537802/job/34614745408?pr=4686) in a follow-up PR.

**Comment by agriyakhetarpal** [comment]
Please feel free to fix the [failing links](https://github.com/pybamm-team/PyBaMM/actions/runs/12399537802/job/34614745408?pr=4686) in this PR or in a quick follow-up, I have no preference.

**Comment by kratman** [comment]
Looking at the failures, it looks like there might be some stuff to fix before we start using this:
```
(source/user_guide/installation/gnu-linux-mac: line   87) broken    index.html#install-required-dependencies - 
(source/user_guide/installation/gnu-linux-mac: line  106) broken    index.html#installation - 
```
It seemed like those links worked when I tried them.

Some of the research paper links fail:
```
(source/api/models/lithium_ion/dfn: line   16) broken    https://www.sciencedirect.com/science/article/pii/S0378775322001604 - 403 Client Error: Forbidden for url: https://www.sciencedirect.com/science/article/pii/S0378775322001604
```

We should probably have fixes in place for those before we merge this, otherwise all PRs will be marked as failing

**Comment by agriyakhetarpal** [comment]
Yeah, just saw the first ones as well, unsure how to fix those... In the meantime, we ignore all links from ScienceDirect with Lychee, we can use a similar option in `conf.py`: https://www.sphinx-doc.org/en/master/usage/configuration.html#confval-linkcheck_ignore.

File: .github/workflows/test_docs_build.yml
Comment (agriyakhetarpal) [inline .github/workflows/test_docs_build.yml:46]:
```suggestion
        run: sphinx-build -b html -W --keep-going -d docs/_build/.doctrees docs docs/_build
```

To let all warnings be interpreted as errors (matching the behaviour of the previous iteration of this check).

**Comment by RohitP2005** [comment]
> Just two more changes, thanks for taking this up, @RohitP2005!

I have fixed those corrections and made a commit.
Thank you @agriyakhetarpal i enjoy working with PyBamm and look forward to contribute more

File: .github/workflows/test_docs_build.yml
Comment (RohitP2005) [reply .github/workflows/test_docs_build.yml]: My latest commit inlcudes all the corrrection that were requested.

Comment (RohitP2005) [reply .github/workflows/test_docs_build.yml]: the "-w" is causing the job to fail, is that fine

**Comment by agriyakhetarpal** [comment]
> > Just two more changes, thanks for taking this up, @RohitP2005!
> 
> I have fixed those corrections and made a commit. Thank you @agriyakhetarpal i enjoy working with PyBamm and look forward to contribute more

Thanks @RohitP2005, though we need the CI to pass before we are able to merge, since there are a lot of failures in the links. Could you investigate how they can be fixed?

To fix the "Git clone too shallow" warning/error, you can add `fetch-depth: 0` as an input to `actions/checkout@v4` in the workflow file you added.

File: .github/workflows/test_docs_build.yml
Comment (agriyakhetarpal) [inline .github/workflows/test_docs_build.yml:23]:
```suggestion
      - name: Check out repository
        uses: actions/checkout@v4
        with:
          fetch-depth: 0
```

**Comment by agriyakhetarpal** [comment]
Actually, I just noticed that Read the Docs supports customising the docs build process before the HTML job as well, which I didn't know about before:

https://docs.readthedocs.io/en/stable/build-customization.html#perform-a-check-for-broken-links

We should drop the extra step on GitHub and move it to our `.readthedocs.yml` file, @RohitP2005 – Read the Docs runs on every commit, and it makes sense to utilise it better, rather than taking away another GHA runner for a longer time than needed.

**Comment by RohitP2005** [comment]
> > > Just two more changes, thanks for taking this up, @RohitP2005!
> > 
> > 
> > I have fixed those corrections and made a commit. Thank you @agriyakhetarpal i enjoy working with PyBamm and look forward to contribute more
> 
> Thanks @RohitP2005, though we need the CI to pass before we are able to merge, since there are a lot of failures in the links. Could you investigate how they can be fixed?
> 
> To fix the "Git clone too shallow" warning/error, you can add `fetch-depth: 0` as an input to `actions/checkout@v4` in the workflow file you added.

yeah @agriyakhetarpal  i was looking into it, as u dicussed it with kratman , I think we should ignore the links from ScienceDirect as a temporary patch .
Also i will look into the Doc support
Does this sound good to you.

**Comment by agriyakhetarpal** [comment]
> I think we should ignore the links from ScienceDirect as a temporary patch .
> Also i will look into the Doc support
> Does this sound good to you.

Yes, sounds good to me.

**Comment by RohitP2005** [comment]
Hey @agriyakhetarpal  
the new commit includes the ignoreing of links from ScienceDirect
and the linkcheck was set to true in ```.readthedocs.yml```

**Review by agriyakhetarpal — CHANGES_REQUESTED** [review]
The build is failing: https://readthedocs.org/projects/pybamm/builds/26674020/

**Comment by RohitP2005** [comment]
@agriyakhetarpal i feel this PR is very clumsy, I will close this PR and open a new one. If its okay with you

**Comment by RohitP2005** [comment]
@agriyakhetarpal i feel this PR is very clumsy, I will close this PR and open a new one

**Pull request closed without merging**
